src/System.Collections/ref/System.Collections.cs

```diff
@@ -140,7 +141,9 @@ public sealed partial class KeyCollection : System.Collections.Generic.ICollecti
             System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator() { throw null; }
             public partial struct Enumerator : System.Collections.Generic.IEnumerator<TKey>, System.Collections.IEnumerator, System.IDisposable
             {
+                private TKey _currentKey;
```

Comment:
Why is this needed when we have both a reference type and a value type field?

Reply:
According to GenAPI's source comments these are the rules followed for compile-time compat:

```
// For compile-time compat, the following rules should work for producing a reference assembly. We drop all private fields, but add back certain synthesized private fields for a value type (struct) as follows:
// - If there are any private fields that are or contain any value type members, add a single private field of type int.
// - And, if there are any private fields that are or contain any reference type members, add a single private field of type object.
// - And, if the type is generic, then for every type parameter of the type, if there are any private fields that are or contain any members whose type is that type parameter, we add a direct private field of that type.
```

The reason for the Generic type fields is to validate that there are no struct layout cycles:
```
// Compiler needs to see any fields, even private, that have generic arguments to be able
// to validate there aren't any struct layout cycles
```